File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cs
using System;
using System.Collections.Generic;

public class Solution {
    static HashSet<int> answer = new HashSet<int>();

    // 중복 제거와 자리 수 조절을 위한 메서드
    static void GetPermutations(char[] arr, int depth, int length)
    {
        if (depth == length)
        {
            string numStr = new string(arr, 0, length);
            int num = int.Parse(numStr);
            CheckPrime(num);
            return;
        }

        for (int i = depth; i < arr.Length; i++)
        {
            Swap(ref arr[depth], ref arr[i]);
            GetPermutations(arr, depth + 1, length);
            Swap(ref arr[depth], ref arr[i]); // backtrack
        }
    }

    static void Swap(ref char a, ref char b)
    {
        char temp = a;
        a = b;
        b = temp;
    }

    static void CheckPrime(int number)
    {
        if (number < 2 || answer.Contains(number)) return;

        for (int i = 2; i * i <= number; i++)
        {
            if (number % i == 0)
                return;
        }

        answer.Add(number);
    }
    
    public int solution(string numbers) {
        answer.Clear();
        char[] digits = numbers.ToCharArray();
        int n = digits.Length;

        for (int i = 1; i <= n; i++)
        {
            GetPermutations(digits, 0, i);
        }

        return answer.Count;
    }
}

File: 프로그래머스/1/12921. 소수 찾기/소수 찾기.cpp
#include <string>
#include <iostream>
#include <vector>

using namespace std;

int solution(int n) {
    int answer = 0;
    vector<bool> is_prime(n + 1, true);
    is_prime[0] = is_prime[1] = false;
    
    for(int i=2;i*i<n;i++){
        if(is_prime[i]){
            for (int j = i*i; j <= n; j += i) {  // i의 배수 지우기 (i*i부터 시작)
                is_prime[j] = false;
            }
        }
    }
    
    for(int i=2;i<n+1;i++){
        if(is_prime[i])
            answer ++;
    }
    
    return answer;
}